automated-testing/performance-testing/readme.md

```diff
@@ -0,0 +1,96 @@
+
+"*Performance testing is a type of testing intended to determine the responsiveness, throughput, reliability, and/or scalability of a system under a given workload.* " -  [Performance Testing Guidance for Web Applications](https://docs.microsoft.com/en-us/archive/blogs/dajung/ebook-pnp-performance-testing-guidance-for-web-applications).
+
+Before getting into the different subcategories of performance tests let us understand why is performance testing typically done.
+
+## Why Performance Testing
+
+Performance testing is commonly conducted to accomplish one or more the following:
+
+* To help in accessing whether a **system is ready for Release**:
+  * Estimating / Predicting the performance characteristics (such as response time, failure rate, throughput) which an application is likely to have when it is released in to production. The results can help in predicting the satisfaction level of the users when interacting with the system. The predicted values can also be compared with agreed values (success criteria) for the performance characteristics when available.
+  * To help in accessing the adequacy of the infrastructure / managed service sku's to meet the desired performance characteristics of a system
+  * Identifying bottlenecks and issues with the application at different load levels
+* To compare the **performance impact of application changes**
+  * Comparing the performance characteristics of an application after a change to the values of performance characteristics during previous runs (or baseline values), can provide an indication of performance issues or enhancements introduced due to a change
+* To **support system tuning**
+  * Comparing performance characteristics of a system for different system configurations
+
+## Key Performance Testing categories
+
+* **Performance Testing** : This category is the super set of all sub categories of performance related testing. It validates/determines the speed, scalability or reliability characteristics of the system under test. Performance testing focusses on achieving the response times, throughput, and resource utilization levels which meet the performance objectives of a system
+* **Load Testing** : This is the subcategory of performance testing which focusses on validating the performance characteristics of a system, when the system faces load volumes which are expected during production operation.
+* **Stress Testing** : This is the subcategory of performance testing which focusses on validating the performance characteristics of a system, when the system faces load volumes or conditions beyond those expected during production operations. This could include tests which limit the memory, disk available to the system.
+
+## Key Performance testing activities
+
+* Understand the [motivation for the tests](#why-performance-testing)
```

Comment:
I have some issues with some of this content.

- ideally, metrics like latency, (sub) linearity with scale are measured during development continuously to ensure that implementation meets targets and that regressions are not introduced.  It is an agile tenet that a system should *always* be ready for release.

- failure rate is not a performance metric -- it is an availability metric

- load testing (i.e. subjecting the target system to volumes of requests) is a means to several ends, including determining capacity limit thresholds, determining costs, measuring phenomenon like transactional fan-out, and maybe others.  I don't see the utility in calling out stress testing as a separate activity from load as defined in the commit.

- missing is the topic of squeeze testing 

- there is a distinction being made between the kinds of tests for the above and end-to-end testing, which IMO is confusing, as typically the kinds of tests under discussion are end-to-end.

Might I suggest a different organization of topics and content?  List the facts that are often needful to be discovered, and for each fact, the activities + techniques that can be applied to determine those facts.  Then folks can pick which facts are pertinent to their engagement and from that what activities therefore need to be included in their game plan.

For instance, if the engagement calls for a net new microservice, then the following facts are interesting:

- does the implementation scale linearly (or sub-linearly) with increasing volume of load wrt latency, resource consumption (CPU, network, mem, RU/s, etc), and cost?  
- what is the cost curve wrt increasing load, in units that are meaningful to operations and business (e.g. cost per million concurrent users)
- what is the melting point of each component in the system, given the production configuration (VM sku, network topology, etc).
- what is the melting point of the system e2e -- which component melts first?
- what are the runbook steps for coping with hitting the melt points?

Reply:
Having said all of that, better to get in some content and iterate than to delay.  So don't take these comments as "refusal to sign off" in any fashion.

Reply:
Once the test type matrix is complete (https://github.com/microsoft/code-with-engineering-playbook/tree/master/automated-testing) it will point to sections for the relevant test type. 

Other subcategories of Performance tests will be added the main performance testing page (Just the summary). Details of the sub category of performance test (like load tests) will be moved to a new page under performance testing.